### ISSUE TITLE
Prematurely abort 'Wait Until Keyword Succeeds' in a programmatic way with a negative response

### ISSUE THREAD

**Issue by joergschultzelutter** [robotframework/robotframework#4077]
(copy of https://forum.robotframework.org/t/wait-until-keyword-succeeds-wuks-and-api-testing/2369 which has a few more detailed explanations on this topic)

I’ve been using WUKS in the context of API testing where the general work flow is as follows:

- Submit an async API transaction and receive an associated transaction ID
- Use WUKS and subsequentially run a query on the transaction ID’s status. This check can take several _hours_ to complete (average setting: 120 retries with a 1 min interval)

The outcome of the WUKS-related keyword can be three-fold:

- __positive outcome__. Exit WUKS and continue with the test. This is our desired outcome.
- __negative outcome__. The test has __failed completely__. See comments below.
- __everything else__ (e.g. async operation is still running). Triggers a "soft" failure which will then execute another WUKS cycle.

When the test knows that its async option has __failed completely__, I want to have the option to terminate the WUKS-associated keyword prematurely in a __negative__ manner, thus terminating the __current__ test __without__ affecting other tests within the same test suite.  There is no sense for the test to wait until it auto-terminates after 120 minutes - at this point, I already know that it has failed and I don't want to keep the user waiting. In order to achieve this premature test exit, I tried the following two attempts:

- __Fail__ keyword. Unsuccessful approach as a __Fail__ will simply trigger another WUKS cycle. In case I have specified a retry interval of 120 times with a sleep cycle of one minute, my test will terminate eventually - but __only__ after 120 minutes.
- __Fatal Error__ keyword. This is the option that is currently part of my WUKS implementation. The moment my test keyword recognises that the WUKS test step has failed completely (and that there is no further sense in waiting until the test auto-terminates), I trigger a __Fatal Error__ and the WUKS keyword and its current test terminates. Works like a charm.

Unfortunately, the __Fatal Error__ keyword does not only terminate the current test but ALL other tests within that same suite. What I’m essentially looking for is a keyword/option that permits me to prematurely end the WUKS cycle with a __negative__ response, thus terminating the __current__ test but __not__ the remainder of the test suite. Read: if test 1's WUKS keyword fails with a negative outcome and the WUKS code was __prematurely__ terminated, suite tests 2..n are still supposed to be executed. 

The only potential option that I was able to come up with is that both __positive__ and __negative__ outcome cases terminate the WUKS test in a positive manner and pass their respective pass/fail status to code that is _outside_ of the WUKS keyword. That outside code then evaluates the status and issues a __Fail__, if necessary. However, this makes the code more complex as each WUKS call requires code that needs to be executed outside of the custom WUKS keyword. Is there a more elegant/recommended approach to this use case that I may have overlooked? 

Pseudo code of the current implementation:

	*** Test Cases ***
	Send API Request
		Send async API request and get the job's processing ID
		Wait Until Keyword Succeeds  120x  1min  MyWUKSKeyword
		Yay, my test was successful!
		
	*** Keywords ***
	MyWUKSKeyword
		Check status of async processing ID
		If test case has failed completely
			Fatal Error --> aborts the WUKS cycle/keyword but should keep the suite running
		Endif
		
		If test result is still incomplete (or non-fatal test case error)
			Fail --> Triggers another WUKS cycle
		Endif
		
		--> End of the WUKS keyword
		--> continue with the test case

**Comment by pekkaklarck** [comment]
I understand what you want to accomplish but I'm not terribly excited to make the WUKS keyword any more complex than it already is. I'm especially worried that this kind of configuration wouldn't be generic enough to solve all similar cases. What I'd consider instead:

1. Implement this logic in a custom library. In your own library you'd have full control on how the logic goes. You wouldn't have a problem that WUKS can increase log size considerably either.
2. Add more programmatic features to Robot's syntax to allow implementing this on Robot side. Basically we'd need `WHILE` loops and `TRY/EXCEPT` functionality. We already have an issue about the latter that's initially targeted for RF 5.0 and the former could possibly be implemented as well.

**Comment by joergschultzelutter** [comment]
Trust me Pekka - I am with you on the 'not terribly excited' part and I rather keep things  simple than to over-complicate them. Unfortunately, my automation use case is really that difficult and I was hoping for some magic wand to make this issue go away with e.g. an assertion error similar to the one that is triggered when the keyword times out. But I can certainly live without it and just wanted to check if there is an option that I may have overlooked.

For now, I will use my 'handle the test case outside of the WUKS keyword' approach. This is doable even without the use of external libraries and just requires me to redesign the test cases that currently use the WUKS approach. Additionally, I already  use the ``removekeywords: WUKS`` option - log file size should not be an issue.

If that approach does not work, I will shift my code towards a custom Python library.

Thanks!

**Comment by pekkaklarck** [comment]
Special exceptions are somewhat heavy weapon and I'd reserve them for more generic and also important cases like SKIP recently. It sounds like custom Python library would be the best solution.

**Comment by joergschultzelutter** [comment]
I am with you on that one, too.

For completeness sake, here is my crude-but-functional RF3 compatible test which seems to cover my use case. 

	*** Settings ***
	Library         Collections

	*** Test Cases ***
	Testfall
		Wait Until Keyword Succeeds  20x  0sec  Wookie
		Run Keyword If  '${WUKS_ABORT_STATUS}' == '${TRUE}'  Fail  msg=${WUKS_ABORT_ERRMSG}

	Noch ein Testfall
		Wait Until Keyword Succeeds  20x  0sec  Wookie
		Run Keyword If  '${WUKS_ABORT_STATUS}' == '${TRUE}'  Fail  msg=${WUKS_ABORT_ERRMSG}

	*** Keywords ***
	Wookie
		Set Global Variable  ${WUKS_ABORT_STATUS}  ${FALSE}
		Set Global Variable  ${WUKS_ABORT_ERRMSG}  ${EMPTY}

		${Nummer}=  Evaluate  random.sample(range(1, 100),1)   random
		${Nummer}=  Get From List  ${Nummer}  0
		Log To Console  ${Nummer}

		# negative case if value == 4
		Run Keyword If          '${Nummer}' == '4'  Set Global Variable  ${WUKS_ABORT_STATUS}  ${True}
		Run Keyword If          '${Nummer}' == '4'  Set Global Variable  ${WUKS_ABORT_ERRMSG}  Scream for me, Chewie!
		Return From Keyword If  '${Nummer}' == '4'

		# positive case if value == 9
		Return From Keyword If   '${Nummer}' == '9'

		# Otherwise, force a new WUKS loop
		Fail

**Comment by guinibir** [comment]
> *** Test Cases ***
> Testfall
> 	Wait Until Keyword Succeeds  20x  0sec  Wookie
> 	Run Keyword If  '${WUKS_ABORT_STATUS}' == '${TRUE}'  Fail  msg=${WUKS_ABORT_ERRMSG}
> 
> Noch ein Testfall
> 	Wait Until Keyword Succeeds  20x  0sec  Wookie
> 	Run Keyword If  '${WUKS_ABORT_STATUS}' == '${TRUE}'  Fail  msg=${WUKS_ABORT_ERRMSG}


in this scenario, the abort status check will occur after the wait until keyword has finished all of its 20 runs, which does not solve the scenario, it still performs all the 20 attempts before giving up
any ideas on how to make it abort prematurely?

**Comment by joergschultzelutter** [comment]
Removing the abort status check from the WUKS code _was_ actually the point which solved my use case. Please let me elaborate. 

If you want to leave the WUKS code, you literally have three options:

- ```Return from keyword``` - exits the WUKS code with a positive return value. We definitely want this at a certain point in time. Alternatively, you can end the WUKS keyword without an error - which will also trigger the same effect.
- ``` Fail``` - triggers a new WUKS cycle (as long as you are within the boundaries of retries specified for the WUKS command). We also want this for as long as our condition has not been met.
- ```Fatal error``` - terminates the entire suite. Obviously, this is undesired behavior in our case.

What I needed was an option where I would be able to exit the WUKS code programmatically _without_ triggering another rerun EVEN IF my WUKS operation had failed. A "positive fail", so to say.

I achieved this approach by setting a global error variable ```${WUKS_ABORT_STATUS}``` to ```True``` _within_ the WUKS code - and then checking on that variable's status _outside_ of the WUKS code. If that variable's value is ```True```, I know that the test has failed - and can then simply kill it off outside of the WUKS code by using the ```Fail``` command.

tl;dr: there are three ways for the code to proceed:
- Option 1: WUKS call exceeds number of reruns --> Robot lets this test fail automatically
- Option 2: We exited the WUKS look with ```${WUKS_ABORT_STATUS}==${FALSE}``` --> ok case; we can continue
- Option 3: We exited the WUKS look with ```${WUKS_ABORT_STATUS}==${TRUE}``` --> we abort the test programmatically.

For the test code, try increasing the ```20x``` to e.g ```200x```. You can  see that the code execution does not happen 200 times - the moment the "good" value has been identified, the code stops.

FWIW, I have this code in production for my API case since 2021 and it's been working flawlessly on a daily basis.  It's not pretty but it definitely works. Feel free to reply in case this approach is unclear.
